protocol/src/protocol.semanticHighlighting.proposed.md

```diff
@@ -0,0 +1,134 @@
+#### Semantic Highlighting
+
+While the syntax highlighting is done on the client-side and can handle keywords, strings, and other low-level tokens from the grammar, it cannot adequately support complex coloring. Semantic highlighting information is calculated on the language server and pushed to the client as a notification. This notification carries information about the ranges that have to be colored. The desired coloring details are given as [TextMate scopes](https://manual.macromates.com/en/language_grammars) for each affected range. For the semantic highlighting information the following additions are proposed:
+
+_Client Capabilities_:
+
+Capability that has to be set by the language client if it can accept and process the semantic highlighting information received from the server.
+
+```ts
+/**
+ * The text document client capabilities.
+ */
+textDocument?: {
+
+	/**
+	 * The client's semantic highlighting capability.
+	 */
+	semanticHighlightingCapabilities?: {
+
+		/**
+		 * `true` if the client supports semantic highlighting support text documents. Otherwise, `false`. It is `false` by default.
+		 */
+		semanticHighlighting: boolean;
+
+	}
+
+}
+```
+
+_Server Capabilities_:
+
+If the client declares its capabilities with respect to the semantic highlighting feature, and if the server supports this feature too, the server should set all the available TextMate scopes as a "lookup table" during the `initialize` request.
+
+```ts
+/**
+ * Semantic highlighting server capabilities.
+ */
+semanticHighlighting?: {
+
+	/**
+	 * A "lookup table" of semantic highlighting [TextMate scopes](https://manual.macromates.com/en/language_grammars)
+	 * supported by the language server. If not defined or empty, then the server does not support the semantic highlighting
+	 * feature. Otherwise, clients should reuse this "lookup table" when receiving semantic highlighting notifications from
+	 * the server.
+	 */
+	scopes?: string[][];
```

Comment:
I had a look at the link in the comment, but I still don't understand what this 2D array is expected to contain.

I assume every top-level element of the array represents one semantic highlighting scope, and the indexes encoded in the `tokens` string index into that top-level array.

If each of those top-level elements is itself an array of strings, what goes into those arrays? The name of the scope, which is one string... and what else?

Perhaps an example would help clarify the intended usage.

Reply:
Ok, I think I shed some light on this by looking at the [Java language server's implementation](https://github.com/eclipse/eclipse.jdt.ls/blob/6819b791a7f5d83e583dd830fd2a7f6948e5e80c/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/highlighting/SemanticHighlightings.java#L69).

IIUC, every semantic highlighting token kind has multiple scopes associated with it, ordered from most specific to least specific. For example, "static method" in Java has:

```
"storage.modifier.static.java", "entity.name.function.java", "meta.function-call.java", "meta.method.body.java", "meta.method.java", "meta.class.body.java", "meta.class.java",
"source.java"
```

That seems like an awful lot of scopes to me, but I guess the general idea is that it allows clients to use both coarse-grained token kinds (e.g. "static method" will match "function", so if a client is configured to highlight functions with a particular color, that will be applied to static methods) and fine-grained token kinds (if a client has a more specific color configured for "static methods", then that will take precedence over the color configured for functions).

Let me know if I'm understanding this correctly.

Reply:
As another implementer on the client side, I also have questions as to how this should be handled. @HighCommander4 when you say 

> every semantic highlighting token kind has multiple scopes associated with it

how does this relate to the definition of a token below? In the Encoding of Tokens section it says that a token has only a single scope index. How does this then map into the 2D space defined here?

When looking at the clangd server side output e.g. I see that the scopes array is 1D in practice anyways:

```
[["variable.other.cpp"],["entity.name.function.cpp"],["entity.name.function.method.cpp"],["variable.other.field.cpp"],["entity.name.type.class.cpp"],["entity.name.type.enum.cpp"],["variable.other.enummember.cpp"],["entity.name.namespace.cpp"],["entity.name.type.template.cpp"]]
```

Does someone know a server side implementation that produces a 2D array?

Reply:
> Does someone know a server side implementation that produces a 2D array?

This [Java server implementation](https://github.com/eclipse/eclipse.jdt.ls/blob/6819b791a7f5d83e583dd830fd2a7f6948e5e80c/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/highlighting/SemanticHighlightings.java#L69) (which I also linked to in the comment that you're replying to) does. The line I link to contains a list of multiple scope strings associated with a single token kind, `staticFinalField`.

> As another implementer on the client side, I also have questions as to how this should be handled. @HighCommander4 when you say
> 
> > every semantic highlighting token kind has multiple scopes associated with it
> 
> how does this relate to the definition of a token below? In the Encoding of Tokens section it says that a token has only a single scope index. How does this then map into the 2D space defined here?

My understanding is that, after indexing into the 2D array with the scope index, the intention is for the client to loop through the resulting 1D array, which is expected to be ordered from "most specific" to "least specific" scope, and use the first one that has a matching style in the user's theme.

So, taking the `staticFinalField` example above, the client would first try the scope `"storage.modifier.static.java"`, and if that matches a style in the user's theme, then apply that. Otherwise, keep looping and try the next scope. This way, even if the user's theme didn't contain a more specific rule (e.g. one for static variables), they'll still get a coloring for a less specific one (e.g. one for variables).